pkg/apis/application/v1alpha1/types_codefresh.go

```diff
@@ -16,3 +16,19 @@ func (a *Application) SetDefaultTypeMeta() {
 		APIVersion: SchemeGroupVersion.String(),
 	}
 }
+
+func (a *ApplicationSpec) GetNonRefSource() (*ApplicationSource, int) {
+	if a.HasMultipleSources() {
+		for idx, source := range a.Sources {
+			if !source.IsRef() {
+				return &source, idx
+			}
+		}
+	}
+
+	if a.Source != nil { // single source app
```

Comment:
Can you please just check if app has source and return this source, otherwise execute logic after. 

It is impossible that app will not have source at all